drivers/clock_control/clock_control_r8a779f_cpg_mssr.c

```diff
@@ -13,72 +13,259 @@
 #include <zephyr/arch/cpu.h>
 #include <zephyr/drivers/clock_control/renesas_cpg_mssr.h>
 #include <zephyr/dt-bindings/clock/renesas_cpg_mssr.h>
+#include <zephyr/dt-bindings/clock/r8a779f_cpg_mssr.h>
 #include <zephyr/irq.h>
 #include "clock_control_renesas_cpg_mssr.h"
 #include <zephyr/logging/log.h>
 
 LOG_MODULE_DECLARE(clock_control_rcar);
 
-/*
- * TODO: add locks to this driver and parent driver, because it can run
- *       on multicore system, for now all boards ue only one CPU, so it
- *       isn't a big problem.
- */
+#define R8A779F_CLK_SD_STOP_BIT BIT(8)
+#define R8A779F_CLK_SD_DIV_MASK 0x3
+#define R8A779F_CLK_SD_DIV_SHIFT 0
+
+#define R8A779F_CLK_SDH_STOP_BIT BIT(9)
+#define R8A779F_CLK_SDH_DIV_MASK 0x7
+#define R8A779F_CLK_SDH_DIV_SHIFT 2
+
+#define R8A779F_CLK_SD0CKCR1_DIV_MASK 0x3
+#define R8A779F_CLK_SD0CKCR1_DIV_SHIFT 29
 
 struct r8a779f_cpg_mssr_cfg {
 	mm_reg_t base_address;
 };
 
+struct r8a779f_cpg_mssr_data {
+	struct rcar_cpg_mssr_data cmn;
+};
+
+/* NOTE: the array MUST be sorted by module field */
+static struct cpg_clk_info_table core_props[] = {
+	RCAR_CORE_CLK_INFO_ITEM(R8A779F_CLK_PLL5VCO, RCAR_CPG_NONE,
+				RCAR_CPG_NONE, RCAR_CPG_MHZ(3200)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A779F_CLK_SD0CKCR1, 0x08A4, R8A779F_CLK_PLL5VCO, RCAR_CPG_NONE),
+	RCAR_CORE_CLK_INFO_ITEM(R8A779F_CLK_SDH, 0x0870, R8A779F_CLK_SD0CKCR1, RCAR_CPG_NONE),
+	RCAR_CORE_CLK_INFO_ITEM(R8A779F_CLK_SD, 0x0870, R8A779F_CLK_SDH, RCAR_CPG_NONE),
+};
+
+/* NOTE: the array MUST be sorted by module field */
+static struct cpg_clk_info_table mod_props[] = {
+	RCAR_MOD_CLK_INFO_ITEM(706, R8A779F_CLK_SD),
+};
+
+
+static int r8a779f_cpg_enable_disable_core(const struct r8a779f_cpg_mssr_cfg *cfg,
+					   struct cpg_clk_info_table *clk_info,
+					   uint32_t enable)
+{
+	int ret = 0;
+	uint32_t reg;
+
+	enable = !!enable;
+
+	switch (clk_info->module) {
+	case R8A779F_CLK_SD0CKCR1:
+		return 0;
+	case R8A779F_CLK_SD:
+		reg = sys_read32(cfg->base_address + clk_info->offset);
+		reg &= ~R8A779F_CLK_SD_STOP_BIT;
+		reg |= !enable * R8A779F_CLK_SD_STOP_BIT;
+		break;
+	case R8A779F_CLK_SDH:
+		reg = sys_read32(cfg->base_address + clk_info->offset);
+		reg &= ~R8A779F_CLK_SDH_STOP_BIT;
+		reg |= !enable * R8A779F_CLK_SDH_STOP_BIT;
+		break;
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
+
+	if (!ret) {
+		rcar_cpg_write(cfg->base_address, clk_info->offset, reg);
+	}
+	return ret;
+}
+
+static int r8a779f_cpg_core_clock_endisable(const struct device *dev,
+					    struct rcar_cpg_clk *clk,
+					    bool enable)
+{
+	struct cpg_clk_info_table *clk_info;
+	const struct r8a779f_cpg_mssr_cfg *cfg = dev->config;
+	struct r8a779f_cpg_mssr_data *data = dev->data;
+	k_spinlock_key_t key;
+	int ret = 0;
+
+	clk_info = rcar_cpg_find_clk_info_by_module_id(dev, clk->domain, clk->module);
+	if (!clk_info) {
+		return -EINVAL;
+	}
+
+	if (enable) {
+		if ((int)clk->rate > 0) {
```

Comment:
`R8A779F_CLK_PLL5VCO` has rate which can't be represented as signed int.

Reply:
Apart from R8A779F0 there are R8A779G0, R8A779A0 and R8A77980. Can you tell the same about theirs clocks?